public/apps/configuration/panels/role-edit/tenant-panel.tsx

```diff
@@ -13,21 +13,123 @@
  *   permissions and limitations under the License.
  */
 
-import React from 'react';
-import { EuiFieldText } from '@elastic/eui';
-import { PanelWithHeader } from '../../utils/panel-with-header';
+import { EuiButton, EuiComboBox, EuiFlexGroup, EuiFlexItem, EuiSuperSelect } from '@elastic/eui';
+import React, { Dispatch, Fragment, SetStateAction } from 'react';
+import { RoleTenantPermission } from '../../types';
+import {
+  appendElementToArray,
+  removeElementFromArray,
+  updateElementInArrayHandler,
+} from '../../utils/array-state-utils';
+import { appendOptionToComboBoxHandler, stringToComboBoxOption } from '../../utils/combo-box-utils';
 import { FormRow } from '../../utils/form-row';
+import { PanelWithHeader } from '../../utils/panel-with-header';
+import { ComboBoxOptions, RoleTenantPermissionStateClass, TenantPermissionType } from './types';
+
+const TENANT_READ_PERMISSION = 'kibana_all_read';
+const TENANT_WRITE_PERMISSION = 'kibana_all_write';
+
+export function buildTenantPermissionState(
+  permissions: RoleTenantPermission[]
+): RoleTenantPermissionStateClass[] {
+  return permissions.map(permission => {
+    const readable = permission.allowed_actions.includes(TENANT_READ_PERMISSION);
+    const writable = permission.allowed_actions.includes(TENANT_WRITE_PERMISSION);
+    let permissionType = TenantPermissionType.None;
+    if (readable && writable) {
+      permissionType = TenantPermissionType.Full;
+    } else if (readable) {
+      permissionType = TenantPermissionType.Read;
+    } else if (writable) {
+      permissionType = TenantPermissionType.Write;
+    }
+    return {
+      tenantPatterns: permission.tenant_patterns.map(stringToComboBoxOption),
+      permissionType,
+    };
+  });
+}
+
+function getEmptyTenantPermission(): RoleTenantPermissionStateClass {
+  return {
+    tenantPatterns: [],
+    permissionType: TenantPermissionType.None,
+  };
+}
+
+function generateTenantPermissionPanels(
+  permissions: RoleTenantPermissionStateClass[],
+  permisionOptionsSet: ComboBoxOptions,
+  setPermissions: Dispatch<SetStateAction<RoleTenantPermissionStateClass[]>>
+) {
+  const panels = permissions.map((permission, arrayIndex) => {
+    const onValueChangeHandler = (attributeToUpdate: string) =>
+      updateElementInArrayHandler(setPermissions, [arrayIndex, attributeToUpdate]);
+
+    const onCreateOptionHandler = (attributeToUpdate: string) =>
+      appendOptionToComboBoxHandler(setPermissions, [arrayIndex, attributeToUpdate]);
 
-export function TenantPanel(props: {}) {
+    return (
+      <Fragment key={`tenant-permission-${arrayIndex}`}>
+        <EuiFlexGroup>
+          <EuiFlexItem style={{ maxWidth: '400px' }}>
+            <EuiComboBox
```

Comment:
Do we prevent the same tenant name from being added multiple times?

Reply:
The combobox does not allow duplicate value. But this didn't check among different rows. Note that the existing version does the same way.

![image](https://user-images.githubusercontent.com/63078162/83207676-a2b80280-a108-11ea-86da-5664e315e85d.png)